lib/validation.js

```diff
@@ -1,39 +1,164 @@
 'use strict'
 
-function validateOpts (opts) {
-  const { all, policy, ttl, cacheSize, onHit, onMiss, onSkip } = opts
+function validateOpts (app, opts = {}) {
+  let { all, policy, ttl, skip, storage, onDedupe, onHit, onMiss, onSkip, onError, logInterval, logReport } = opts
 
   if (all && typeof all !== 'boolean') {
-    throw new Error('all must be an boolean')
-  }
-
-  if (policy && typeof policy !== 'object' && !all) {
-    throw new Error('policy must be an object')
+    throw new Error('all must be a boolean')
   }
 
   if (all && policy) {
     throw new Error('policy and all options are exclusive')
   }
 
-  if (ttl && typeof ttl !== 'number') {
-    throw new Error('ttl must be a number')
+  if (ttl && (typeof ttl !== 'number' || ttl < 0)) {
+    throw new Error('ttl must be a number greater than 0')
+  }
+
+  if (skip && typeof skip !== 'function') {
+    throw new Error('skip must be a function')
+  }
+
+  if (onDedupe) {
+    if (typeof onDedupe !== 'function') {
+      throw new Error('onDedupe must be a function')
+    }
+  } else {
+    onDedupe = noop
+  }
+
+  // TODO move report options validation to report lib
+  if (logInterval && (typeof logInterval !== 'number' || logInterval < 1)) {
+    throw new Error('logInterval must be a number greater than 1')
+  }
+
+  if (logReport && typeof logReport !== 'function') {
+    throw new Error('logReport must be a function')
+  }
+
+  if (onHit) {
+    if (typeof onHit !== 'function') {
+      throw new Error('onHit must be a function')
+    }
+  } else {
+    onHit = noop
+  }
+
+  if (onMiss) {
+    if (typeof onMiss !== 'function') {
+      throw new Error('onMiss must be a function')
+    }
+  } else {
+    onMiss = noop
   }
 
-  if (cacheSize && typeof cacheSize !== 'number') {
-    throw new Error('cacheSize must be a number')
+  if (onSkip) {
+    if (typeof onSkip !== 'function') {
+      throw new Error('onSkip must be a function')
+    }
+  } else {
+    onSkip = noop
   }
 
-  if (onHit && typeof onHit !== 'function') {
-    throw new Error('onHit must be a function')
+  if (onError) {
+    if (typeof onError !== 'function') {
+      throw new Error('onError must be a function')
+    }
+  } else {
+    onError = noop
```

Comment:
Let's change this default implementation. It would significantly better if by default we logged this error as `debug`.